mobile/lib/shared/services/hash.service.dart

```diff
@@ -0,0 +1,183 @@
+import 'package:immich_mobile/shared/models/ios_device_asset.dart';
+import 'package:immich_mobile/shared/providers/db.provider.dart';
+import 'package:immich_mobile/utils/builtin_extensions.dart';
+import 'package:isar/isar.dart';
+import 'package:logging/logging.dart';
+import 'package:photo_manager/photo_manager.dart';
+
+class HashService {
+  HashService(this._db, this._backgroundService);
+  final Isar _db;
+  final BackgroundService _backgroundService;
+  final _log = Logger('HashService');
+
+  Future<List<Asset>> getHashedAssets(
+    AssetPathEntity album, {
+    int start = 0,
+    int end = 0x7fffffffffffffff,
+    Set<String>? excludedAssets,
+  }) async {
+    final entities = await album.getAssetListRange(start: start, end: end);
+    final filtered = excludedAssets == null
+        ? entities
+        : entities.where((e) => !excludedAssets.contains(e.id)).toList();
+    return hashAssets(filtered);
+  }
+
+  Future<List<int>?> hashAsset(Asset a) async {
+    assert(a.isLocal, "can only get hash of local assets");
+    final deviceAsset = await (Platform.isAndroid
+        ? _db.androidDeviceAssets.get(a.localId!.toInt())
+        : _db.iOSDeviceAssets.getById(a.localId!));
+    if (deviceAsset != null) return deviceAsset.hash;
+    final file = await a.local!.originFile;
+    if (file == null) {
+      _log.warning("Failed to get file for asset ${a.id}, skipping");
+      return null;
+    }
+    final hash =
+        await (Platform.isAndroid ? _hashFile(file) : _hashAssetCrypto(file));
+    if (hash == null) {
+      _log.warning("Failed to hash file ${file.path} for asset ${a.id}");
+      return null;
+    }
+    if (Platform.isAndroid) {
+      final da = AndroidDeviceAsset(id: a.localId!.toInt(), hash: hash);
+      await _db.writeTxn(() => _db.androidDeviceAssets.put(da));
+    } else {
+      final da = IOSDeviceAsset(id: a.localId!, hash: hash);
+      await _db.writeTxn(() => _db.iOSDeviceAssets.put(da));
+    }
+    return hash;
+  }
+
+  Future<List<Asset>> hashAssets(List<AssetEntity> assets) async {
+    final Stopwatch totalTime = Stopwatch()..start();
+    const int batchFileCount = 128;
+    const int batchDataSize = 1024 * 1024 * 1024; // 1GB
+    int totalBytes = 0;
+    int totalFiles = 0;
+    final List<DeviceAsset?> hashes;
+    final Stopwatch sw = Stopwatch()..start();
+    final List<DeviceAsset> toAdd = [];
+    final ids = assets
+        .map(Platform.isAndroid ? (a) => a.id.toInt() : (a) => a.id)
+        .toList();
+    hashes = await (Platform.isAndroid
+        ? _db.androidDeviceAssets.getAll(ids.cast())
+        : _db.iOSDeviceAssets.getAllById(ids.cast()));
+    debugPrint("Checking hash DB took ${sw.elapsedMilliseconds}ms");
+    sw.reset();
+    int bytes = 0;
+    final List<String> toHash = [];
+    for (int i = 0; i < assets.length; i++) {
+      if (hashes[i] != null) continue;
+      final file = await assets[i].originFile;
+      if (file == null) {
```

Comment:
What if the last file in this list is `null`? Does the last queue of `toHash` never get processed?